component/stream/stream_cache.go

```diff
@@ -200,3 +210,104 @@ func (c *cache) filePurge(key, value interface{}) {
 	c.blocks.Remove(key)
 	c.evictedBlock = blockKey{}
 }
+
+// Using key construct a file name for persisted block
+func (c *cache) getLocalFilePath(key blockKey) string {
+	return filepath.Join(c.diskPath, key.fileKey+"__"+fmt.Sprintf("%d", key.offset)+"__")
+}
+
+// Persist this block on disk
+func (c *cache) persistBlockOnDisk(block *cacheBlock, key blockKey) {
+	localPath := c.getLocalFilePath(key)
+
+	log.Debug("streamcache::persistBlockOnDisk : Saving file %s offset %d to disk", key.fileKey, key.offset)
+
+	err := os.MkdirAll(filepath.Dir(localPath), os.FileMode(0775))
+	if err != nil {
+		log.Err("streamcache::persistBlockOnDisk : unable to create local directory %s [%s]", localPath, err.Error())
+		return
+	}
+
+	f, err := os.OpenFile(localPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0775)
+	if err != nil {
+		log.Err("streamcache::persistBlockOnDisk : Failed to create file for persisting block on disk %s (%s)", localPath, err.Error())
+		return
+	}
+	f.Write(block.data)
+	f.Close()
+
+	c.diskBlocks.Set(key, true)
+}
+
+// Load data for this block from disk
+func (c *cache) getBlockFromDisk(block *cacheBlock, key blockKey) bool {
+	localPath := c.getLocalFilePath(key)
+	info, err := os.Stat(localPath)
+
+	if err != nil {
+		return false
+	}
+
+	log.Debug("streamcache::getBlockFromDisk : Reading block for %s offset %d from disk", key.fileKey, key.offset)
+	if time.Since(info.ModTime()).Seconds() > c.diskTimeoutSec {
+		// File exists on local disk but disk cache timeout has elapsed
+		os.Remove(localPath)
+		c.diskBlocks.Remove(key)
+		return false
+	}
+
+	f, err := os.OpenFile(localPath, os.O_RDONLY, 0775)
+	if err != nil {
+		return false
+	}
+
+	f.Read(block.data)
+	f.Close()
+	os.Remove(localPath)
+
+	c.diskBlocks.Remove(key)
```

Comment:
We should leave a comment here explaining:
```suggestion
        // we remove the block from disk as it is now in our in-memory cache
	c.diskBlocks.Remove(key)
```

Reply:
done